bin/propolis-server/src/lib/vm/objects.rs

```diff
@@ -0,0 +1,497 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A collection of all of the components that make up a Propolis VM instance.
+
+use std::{
+    ops::{Deref, DerefMut},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use futures::{future::BoxFuture, stream::FuturesUnordered, StreamExt};
+use propolis::{
+    hw::{ps2::ctrl::PS2Ctrl, qemu::ramfb::RamFb, uart::LpcUart},
+    vmm::VmmHdl,
+    Machine,
+};
+use propolis_api_types::instance_spec::v0::InstanceSpecV0;
+use slog::{error, info};
+use tokio::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use crate::{serial::Serial, vcpu_tasks::VcpuTaskController};
+
+use super::{
+    state_driver::VmStartReason, BlockBackendMap, CrucibleBackendMap, DeviceMap,
+};
+
+/// A collection of components that make up a Propolis VM instance.
+pub(crate) struct VmObjects {
+    /// The objects' associated logger.
+    log: slog::Logger,
+
+    /// A reference to the VM state machine that created these objects. Used to
+    /// complete rundown when the objects are dropped.
+    parent: Arc<super::Vm>,
+
+    /// Synchronizes access to the VM's objects.
+    ///
+    /// API-layer callers that want to enumerate a VM's devices or read its spec
+    /// acquire this lock shared. The state driver acquires this lock exclusive
+    /// to mutate the VM.
+    inner: RwLock<VmObjectsLocked>,
+}
+
+/// A collection of objects that should eventually be wrapped in a lock and
+/// stored in a `VmObjects` structure. See [`VmObjectsLocked`].
+pub(super) struct InputVmObjects {
+    pub instance_spec: InstanceSpecV0,
+    pub vcpu_tasks: Box<dyn VcpuTaskController>,
+    pub machine: Machine,
+    pub lifecycle_components: DeviceMap,
```

Comment:
Since we're keeping the `DeviceMap` name for now, would we want something to reflect that here as well, rather than `lifecycle_components`?  Up to you.

Reply:
Done in 8606927.